tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -6,129 +6,548 @@
+
+            var tracestate = CreateTraceStateHeader(context);
+
+            if (!string.IsNullOrWhiteSpace(tracestate))
+            {
+                carrierSetter.Set(carrier, TraceStateHeaderName, tracestate);
+            }
+        }
+
+        internal static string CreateTraceParentHeader(SpanContext context)
+        {
+            var traceId = IsValidHexString(context.RawTraceId, length: 32) ? context.RawTraceId : context.TraceId.ToString("x32");
+            var spanId = IsValidHexString(context.RawSpanId, length: 16) ? context.RawSpanId : context.SpanId.ToString("x16");
+            var samplingPriority = context.TraceContext?.SamplingPriority ?? context.SamplingPriority ?? SamplingPriorityValues.AutoKeep;
             var sampled = samplingPriority > 0 ? "01" : "00";
-            carrierSetter.Set(carrier, TraceParent, $"00-{traceId}-{spanId}-{sampled}");
+
+            return $"00-{traceId}-{spanId}-{sampled}";
         }
 
-        public bool TryExtract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter, out SpanContext? spanContext)
-            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        internal static string CreateTraceStateHeader(SpanContext context)
         {
-            spanContext = null;
+            var samplingPriority = SamplingPriorityToString(context.TraceContext?.SamplingPriority ?? context.SamplingPriority);
+            StringBuilder? sb = null;
 
-            var traceParent = ParseUtility.ParseString(carrier, carrierGetter, TraceParent)?.Trim();
-            if (!string.IsNullOrEmpty(traceParent))
+            try
             {
-                // We found a trace parent (we are reading from the Http Headers)
+                sb = StringBuilderCache.Acquire(100);
+                sb.Append("dd=");
+
+                if (samplingPriority != null)
+                {
+                    sb.Append("s:").Append(samplingPriority).Append(';');
+                }
 
-                /* (https://www.w3.org/TR/trace-context/)
-                    Valid traceparent when caller sampled this request:
+                if (!string.IsNullOrWhiteSpace(context.Origin))
+                {
+                    sb.Append("o:").Append(context.Origin).Append(';');
+                }
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 01  // sampled
-                    Valid traceparent when caller didn’t sample this request:
+                if (context.TraceContext?.Tags?.ToArray() is { Length: > 0 } tags)
+                {
+                    foreach (var tag in tags)
+                    {
+                        if (tag.Key.StartsWith("_dd.p.", StringComparison.Ordinal))
```

Comment:
not related to this PR (and so clearly not asking for change), but I wonder if we should have a `PropagatedTags` instance in TraceContext to avoid checking the `_dd.p.` in other places. It feels like an internal detail of the propagation that we are using in a few places

Reply:
We check for `_dd.p.*` because `TraceContext.Tags` was designed to hold all "trace-level" tags, but not all of them are necessarily _propagated_ tags.

I do want to come back later and clean that up, though. For example, there are some optimizations in `TraceTagCollection` that expect `_dd.p.*` but that was designed for `x-datadog-tags`. One of them is that we cache the header value as long as the propagated tags don't change to avoid creating it multiple times, but we can't use that cached value for `tracestate`.

For what it's worth, you made me realize I can replace all the `"_dd.p."` in this file with `TagPropagation.PropagatedTagPrefix`, so I did that now. Thanks!